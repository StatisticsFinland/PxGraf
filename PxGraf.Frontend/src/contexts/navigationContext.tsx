import * as React from "react";

interface INavigationContextData {
    tablePath: string[] | null,
}

interface INavigationContextMethods {
    setTablePath: React.Dispatch<React.SetStateAction<string[] | null>>,
}

type TNavigationContextValue = INavigationContextData & INavigationContextMethods;

interface INavigationProviderProps {
    children: React.ReactNode;
}

const NavigationContext = React.createContext<(TNavigationContextValue)>(null);

export const NavigationProvider: React.FC<INavigationProviderProps> = ({ children }) => {
    const [tablePath, setTablePath] = React.useState<string[] | null>(null);

    const contextValue: TNavigationContextValue = React.useMemo(() => ({
        tablePath,
        setTablePath,
    }), [tablePath]);

    return <NavigationContext.Provider value={contextValue}>{children}</NavigationContext.Provider>
}

export const useNavigationContext = (): TNavigationContextValue => {
    return React.useContext(NavigationContext);
}
