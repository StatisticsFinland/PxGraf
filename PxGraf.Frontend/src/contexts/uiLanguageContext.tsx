/* istanbul ignore file */

import * as React from 'react';
import { useTranslation } from 'react-i18next';

/**
 * Interface for UI language context properties
 * @property language - The current UI language. These languages are defined in ../localization/translationsConfig.json.
 * @property setLanguage - Function to set the UI language
 * @property languageTab - The current content language. These languages are defined by the content retrieved from the API and are used to customize the visualization.
 * @property setLanguageTab - Function to set the content language
 * @property availableUiLanguages - List of available UI languages
 * @property uiContentLanguage - The current UI content language. These languages are defined by the content retrieved from the API, but are used in the UI.
 * @property setUiContentLanguage - Function to set the UI content language
 */
interface IUiLanguageContext {
    language: string;
    setLanguage: React.Dispatch<string>;
    languageTab: string;
    setLanguageTab: React.Dispatch<string>;
    availableUiLanguages: string[];
    uiContentLanguage: string;
    setUiContentLanguage: React.Dispatch<string>;
}

export const UiLanguageContext = React.createContext<IUiLanguageContext>({
    language: "",
    setLanguage: () => { /* no base implementation */ },
    languageTab: '',
    setLanguageTab: () => { /* no base implementation */ },
    availableUiLanguages: [],
    uiContentLanguage: '',
    setUiContentLanguage: () => { /* no base implementation */ }
});

/**
 * Provider for the UI language context
 * @param children - The child components
 */
export const UiLanguageProvider: React.FC<{ children: React.ReactNode }> = ({ children }) => {
    const { i18n } = useTranslation();
    const [language, setLanguage] = React.useState(i18n.language);
    const [languageTab, setLanguageTab] = React.useState(i18n.language);
    const [availableUiLanguages, setAvailableUiLanguages] = React.useState<string[]>([]);
    const [uiContentLanguage, setUiContentLanguage] = React.useState(i18n.language);

    const loadLanguages = async () => {
        // Load available languages from i18n
        const languages = Object.keys(i18n.services.resourceStore.data);
        setAvailableUiLanguages(languages);
        const defaultLanguage = languages.includes(i18n.language) ? i18n.language : languages[0];
        if (language !== defaultLanguage) {
            setLanguage(defaultLanguage);
        }
    }

    React.useEffect(() => {
        loadLanguages();
    }, []);

    const setUiLang = (lang: string) => {
        i18n.changeLanguage(lang);
        setLanguage(lang);
    }

    const contextValue = React.useMemo(() => {
        return { language, setLanguage: setUiLang, languageTab, setLanguageTab, availableUiLanguages, uiContentLanguage, setUiContentLanguage }
    }, [language, languageTab, availableUiLanguages, uiContentLanguage]);

    return (
        <UiLanguageContext.Provider value={contextValue}>
            {children}
        </UiLanguageContext.Provider>
    );
}

export default UiLanguageContext;
