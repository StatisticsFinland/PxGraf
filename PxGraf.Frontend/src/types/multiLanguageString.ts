/* istanbul ignore file */

/**
 * Type for string supporting multiple languages with each key representing a language and the value the string in that language.
 */
export type MultiLanguageString = { [key: string]: string };