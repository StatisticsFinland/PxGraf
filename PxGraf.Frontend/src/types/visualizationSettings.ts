/* istanbul ignore file */

import {VisualizationType} from './visualizationType';

/**
 * Interface for visualization settings
 * @property {string[]} defaultSelectableVariableCodes - Dimension codes for default selectable dimension values.
 * @property {boolean} pivotRequested - Flag to indicate if the user has requested manual pivoting.
 * @property {boolean} cutYAxis - Flag to indicate if the y-axis should be cut based on the values.
 * @property {string} multiselectableVariableCode - The dimension code for the multiselectable dimension if any.
 * @property {string[]} rowVariableCodes - Dimension codes for the row dimensions.
 * @property {string[]} columnVariableCodes - Dimension codes for the column dimensions.
 * @property {string} sorting - Selected sorting order.
 * @property {boolean} matchXLabelsToEnd - Flag to indicate if the x-axis labels should be matched to the end.
 * @property {number} markerSize - The size of scatter plot markers.
 * @property {VisualizationType} selectedVisualization - The selected visualization type.
 * @property {boolean} showDataPoints - Flag to indicate if data points should be shown in the visualization.
 */
export interface IVisualizationSettings {
    defaultSelectableVariableCodes?: { [key: string]: string[] };
    pivotRequested?: boolean;
    cutYAxis?: boolean;
    multiselectableVariableCode?: string;
    rowVariableCodes?: string[];
    columnVariableCodes?: string[];
    sorting?: string;
    matchXLabelsToEnd?: boolean;
    markerSize?: number;
    selectedVisualization?: VisualizationType;
    showDataPoints?: boolean;
}