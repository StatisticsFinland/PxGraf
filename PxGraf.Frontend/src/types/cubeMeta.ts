/* istanbul ignore file */

import { MultiLanguageString } from "./multiLanguageString";

/**
 * Interface for cube meta properties. A cube is a multi-dimensional result of Px file data filtered by a query.
 * @property {string[]} languages - List of available languages.
 * @property {MultiLanguageString} header - The multi language header of the cube.
 * @property {MultiLanguageString} note - The note of the cube.
 * @property {IVariable[]} variables - List of variables.
 */
export interface IMatrixMetadata {
    DefaultLanguage: string,
    AvailableLanguages: string[],
    Dimensions: IDimension[],
    AdditionalProperties?: { [key: string]: IMetaProperty }
}

/**
 * TODO: Summary
 */
export interface IMetaProperty {
    KeyWord: string,
    CanGetStringValue: boolean,
    CanGetMultilanguageValue: boolean,
    Entries: MultiLanguageString | string
}

/**
 * Interface for dimnension properties. Each variable defines a dimension in a cube.
 * @property {string} code - The code of the variable.
 * @property {VariableType} type - The type of the variable.
 * @property {MultiLanguageString} name - The multi language name of the variable.
 * @property {{ [key: string]: IMetaProperty }} additionalProperties - Optional dictionary for additional metadata properties
 * @property {IVariableValue[]} values - List of values associated with this variable.
 */
export interface IDimension {
    Code: string,
    Name: MultiLanguageString,
    AdditionalProperties?: { [key: string]: IMetaProperty },
    Values: IDimensionValue[]
    Type: VariableType,
}

/**
 * Interface for variable properties.Each variable defines a dimension in a cube.
 * @property { string } code - The code of the variable.
 * @property { MultiLanguageString } name - The multi language name of the variable.
 * @property { MultiLanguageString } note - The multi language note of the variable.
 * @property { VariableType } type - The type of the variable.
 * @property { IVariableValue[] } values - List of values associated with this variable.
 */
export interface IVariable {
    code: string,
    name: MultiLanguageString,
    note: MultiLanguageString,
    type: VariableType,
    values: IVariableValue[]
}

/**
 * Interface for dimension value properties.
 * @property {string} code - The code of the value.
 * @property {MultiLanguageString} name - The multi language name of the value.
 * @property {{ [key: string]: IMetaProperty }} additionalProperties - Optional dictionary for additional metadata properties
 */
export interface IDimensionValue {
    Code: string,
    Name: MultiLanguageString,
    Virtual: boolean
    AdditionalProperties?: { [key: string]: IMetaProperty },
}

/**
 * Interface for variable value properties.
 * @property { string } code - The code of the value.
 * @property { MultiLanguageString } name - The multi language name of the value.
 * @property { MultiLanguageString } note - The multi language note of the value.
 * @property { boolean } isSum - Flag to indicate if the value is an elimination value.
 * @property { IContentComponent } contentComponent - The content variable component of the value.
 * @property { IVirtualComponent[] } virtualComponents - List of virtual components.
 */
export interface IVariableValue {
    code: string,
    name: MultiLanguageString,
    note: MultiLanguageString,
    isSum: boolean,
    contentComponent?: IContentComponent
    virtualComponents?: IVirtualComponent[]
}

/**
 * Interface for value content variable value properties.
 * @property {MultiLanguageString} unit - The unit of the value.
 * @property {MultiLanguageString} source - The source of the value.
 * @property {number} numberOfDecimals - The precision, number of decimals of the value.
 * @property {string} lastUpdated - The last updated date of the value.
 */
export interface IContentComponent {
    unit: MultiLanguageString,
    source: MultiLanguageString,
    numberOfDecimals: number,
    lastUpdated: string
}

/**
 * Interface for virtual component properties.
 * @property {string} operator - The operator of the virtual component.
 * @property {string[]} operandCoded - List of coded operands.
 */
export interface IVirtualComponent {
    operator: string,
    operandCoded: string[]
}

/**
 * Enumeration for variable types.
 */
export enum VariableType {
    Unknown = 'N',
    Time = 'T',
    Ordinal = 'P',
    Geographical = 'G',
    Content = 'C',
    OtherClassificatory = 'F',
}

export function getVariablesFromMatrixMetadata(meta: IMatrixMetadata): IVariable[] {
    return meta.Dimensions.map(d => {
        return {
            code: d.Code,
            name: d.Name,
            note: d.Name,
            type: d.Type,
            values: d.Values.map(v => convertDimensionValueToVariableValue(v, meta))
        }
    })
}

export function convertDimensionValueToVariableValue(value: IDimensionValue, meta: IMatrixMetadata): IVariableValue {
    const dimension: IDimension = meta.Dimensions.find(d => d.Values.find(v => v.Code === value.Code));
    const contentComponent = (): IContentComponent | null => {
        if (dimension.Type === VariableType.Content) {
            return {
                unit: getAdditionalProperty("UNIT", value.AdditionalProperties) as MultiLanguageString,
                source: getAdditionalProperty("SOURCE", value.AdditionalProperties) as MultiLanguageString,
                numberOfDecimals: parseInt(getAdditionalProperty("PRECISION", value.AdditionalProperties, true) as string),
                lastUpdated: getAdditionalProperty("LAST-UPDATED", value.AdditionalProperties, true) as string
            }
        }
        else return null;
    }
    return {
        code: value.Code,
        name: value.Name,
        note: value.Name,
        isSum: getValueIsSumValue(value, dimension),
        contentComponent: contentComponent()
    }
}

export function getValueIsSumValue(value: IDimensionValue, dimension: IDimension): boolean {
    return getAdditionalProperty("ELIMINATION", dimension.AdditionalProperties) === value.Code;
}

function getAdditionalProperty(key: string, additionalProperties: { [key: string]: IMetaProperty }, forceToString = false): string | MultiLanguageString {
    if (additionalProperties[key]) {
        if (additionalProperties[key].CanGetStringValue) {
            return additionalProperties[key].Entries[0];
        }
        else if (additionalProperties[key].CanGetMultilanguageValue) {
            if (forceToString) return additionalProperties[key].Entries[0];
            else return additionalProperties[key].Entries;
        }
    }
    return null;
}
