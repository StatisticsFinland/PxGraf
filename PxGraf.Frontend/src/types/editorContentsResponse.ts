import { MultiLanguageString } from "./multiLanguageString";
import { VisualizationType } from "./visualizationType";

/**
 * Interface for dimension sorting options.
 * @property {string} code - The code of the sorting option.
 * @property {MultiLanguageString} description - Longer, user facing description of the sorting option.
 */
export interface ISortingOption {
    code: string;
    description: MultiLanguageString;
}

/**
 * Interface for sorting option collection. Contains default and pivoted sorting options if available.
 * @property {ISortingOption[]} default - Default sorting options.
 * @property {ISortingOption[]} pivoted - Pivoted sorting options.
 */
interface ISortingOptionCollection {
    default: ISortingOption[];
    pivoted: ISortingOption[];
}

/**
 * Interface for visualization options for a specific visualization type.
 * @property {VisualizationType} type - The type of visualization.
 * @property {boolean} [allowShowingDataPoints] - Flag to indicate if showing data points is allowed.
 * @property {boolean} [allowCuttingYAxis] - Flag to indicate if cutting the Y-axis is allowed.
 * @property {boolean} [allowMatchXLabelsToEnd] - Flag to indicate if matching X labels to the end is allowed.
 * @property {boolean} [allowSetMarkerScale] - Flag to indicate if setting marker scale is allowed.
 * @property {boolean} allowManualPivot - Flag to indicate if manual pivoting is allowed.
 * @property {boolean} allowMultiselect - Flag to indicate if multivalue selectable is allowed.
 * @property {ISortingOptionCollection} sortingOptions - Sorting options for the visualization containing default and pivoted options.
 */
export interface IVisualizationOptions {
    type: VisualizationType;
    allowShowingDataPoints?: boolean;
    allowCuttingYAxis?: boolean;
    allowMatchXLabelsToEnd?: boolean;
    allowSetMarkerScale?: boolean;
    allowManualPivot: boolean;
    allowMultiselect: boolean;
    sortingOptions: ISortingOptionCollection;
}

/**
 * Interface for the response from the editor contents API used to populate the settings for adjusting the visualization in the editor.
 * @property {number} size - The size of the response based on the amount of values selected from all dimensions.
 * @property {number} sizeWarningLimit - The size warning limit for the response.
 * @property {number} maximumSupportedSize - The maximum supported size for the response.
 * @property {MultiLanguageString} headerText - Default header text for the visualization as a @ {@link MultiLanguageString} object.
 * @property {number} maximumHeaderLength - The maximum length of the header text.
 * @property {IVisualizationOptions[]} visualizationOptions - Visualization options for the valid visualization types as an array of @ {@link IVisualizationOptions} objects.
 * @property {object} visualizationRejectionReasons - Object containing rejection reasons for each visualization type as a @ {@link MultiLanguageString} object.
 */
export interface IEditorContentsResponse {
    size: number;
    sizeWarningLimit: number;
    maximumSupportedSize: number;
    headerText: MultiLanguageString;
    maximumHeaderLength: number;
    visualizationOptions: IVisualizationOptions[];
    visualizationRejectionReasons: {
        [visualizationType: string]: MultiLanguageString;
    }
}