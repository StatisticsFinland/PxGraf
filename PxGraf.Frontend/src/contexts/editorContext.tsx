/* istanbul ignore file */

import { debounce } from 'lodash';
import * as React from 'react';
import { ICubeQuery, Query } from 'types/query';
import { IVisualizationSettings } from 'types/visualizationSettings';
import { VisualizationType } from 'types/visualizationType';

/**
 * Interface for editor context properties
 * @property {ICubeQuery} cubeQuery - Query object that contains the header and dimension queries
 * @property {React.Dispatch<ICubeQuery>} setCubeQuery - Function to set the cube query
 * @property {Query} query - Query object that contains the dimension queries
 * @property {React.Dispatch<Query>} setQuery - Function to set the query
 * @property {boolean} saveDialogOpen - Flag to indicate if the save dialog is open
 * @property {React.Dispatch<boolean>} setSaveDialogOpen - Function to set the save dialog open flag
 * @property {VisualizationType} selectedVisualizationUserInput - The selected visualization type
 * @property {React.Dispatch<VisualizationType>} setSelectedVisualizationUserInput - Function to set the selected visualization type
 * @property {IVisualizationSettings} visualizationSettingsUserInput - The visualization settings
 * @property {React.Dispatch<IVisualizationSettings>} setVisualizationSettingsUserInput - Function to set the visualization settings
 * @property {{ [key: string]: string[] }} defaultSelectables - The default selectables for each dimension
 * @property {React.Dispatch<{ [key: string]: string[] }>} setDefaultSelectables - Function to set the default selectables
 */
interface IEditorContext {
    cubeQuery: ICubeQuery;
    setCubeQuery: React.Dispatch<ICubeQuery>;
    query: Query;
    setQuery: React.Dispatch<Query>;
    saveDialogOpen: boolean;
    setSaveDialogOpen: React.Dispatch<boolean>;
    selectedVisualizationUserInput: VisualizationType;
    setSelectedVisualizationUserInput: React.Dispatch<VisualizationType>;
    visualizationSettingsUserInput: IVisualizationSettings;
    setVisualizationSettingsUserInput: React.Dispatch<IVisualizationSettings>;
    defaultSelectables: { [key: string]: string[] };
    setDefaultSelectables: React.Dispatch<{ [key: string]: string[] }>
}

/**
 * Context required for the editor
 */
export const EditorContext = React.createContext<IEditorContext>({
    cubeQuery: null,
    setCubeQuery: () => { /* no base implementation */ },
    query: null,
    setQuery: () => { /* no base implementation */ },
    saveDialogOpen: false,
    setSaveDialogOpen: () => { /* no base implementation */ },
    selectedVisualizationUserInput: null,
    setSelectedVisualizationUserInput: () => { /* no base implementation */ },
    visualizationSettingsUserInput: null,
    setVisualizationSettingsUserInput: () => { /* no base implementation */ },
    defaultSelectables: null,
    setDefaultSelectables: () => { /* no base implementation */ }
});

/**
 * Provider for the editor context
 * @param children - The child components
 */
export const EditorProvider: React.FC<{ children: React.ReactNode }> = ({ children }) => {
    const [cubeQuery, setCubeQueryState] = React.useState({ variableQueries: {} });
    const [query, setQuery] = React.useState(null);
    const [saveDialogOpen, setSaveDialogOpen] = React.useState(false);
    const [selectedVisualizationUserInput, setSelectedVisualizationUserInput] = React.useState(null);
    const [visualizationSettingsUserInput, setVisualizationSettingsUserInput] = React.useState(null);
    const [defaultSelectables, setDefaultSelectables] = React.useState(null);

    const debouncedSetCubeQuery = React.useMemo(() => {
        return debounce((newQuery: ICubeQuery) => {
            setCubeQueryState(newQuery);
        }, 1000);
    }, []);

    const setCubeQuery = React.useCallback((newQuery: ICubeQuery) => {
        debouncedSetCubeQuery(newQuery);
    }, [debouncedSetCubeQuery]);

    React.useEffect(() => {
        return () => {
            debouncedSetCubeQuery.cancel();
        };
    }, [debouncedSetCubeQuery]);

    const contextValue = React.useMemo(() => {
        return {
            cubeQuery, setCubeQuery,
            query, setQuery,
            saveDialogOpen, setSaveDialogOpen,
            selectedVisualizationUserInput, setSelectedVisualizationUserInput,
            visualizationSettingsUserInput, setVisualizationSettingsUserInput,
            defaultSelectables, setDefaultSelectables
        }
    }, [
        cubeQuery, query, saveDialogOpen, selectedVisualizationUserInput,
        visualizationSettingsUserInput, defaultSelectables
    ]);

    return (
        <EditorContext.Provider value={contextValue}>
            {children}
        </EditorContext.Provider>
    );
}