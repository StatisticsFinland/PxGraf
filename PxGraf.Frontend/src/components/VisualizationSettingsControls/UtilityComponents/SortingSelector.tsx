import {
    FormControl, Select, MenuItem, InputLabel
} from '@mui/material';
import React from 'react';
import { useTranslation } from 'react-i18next';
import { UiLanguageContext } from 'contexts/uiLanguageContext';
import { ISortingOption } from 'types/visualizationRules';
import { IVisualizationSettings } from '../../../types/visualizationSettings';
import { EditorContext } from '../../../contexts/editorContext';

interface ISortingSelectorProps {
    sortingOptions: ISortingOption[],
    visualizationSettings: IVisualizationSettings
}

export const SortingSelector: React.FC<ISortingSelectorProps> = ({ sortingOptions, visualizationSettings }) => {
    const { t } = useTranslation();
    const { uiContentLanguage } = React.useContext(UiLanguageContext);
    const { setVisualizationSettingsUserInput } = React.useContext(EditorContext);

    return (
        <FormControl>
            <InputLabel id="sorting-selector-label">{t("chartSettings.sort")}</InputLabel>
            <Select
                sx={{ minWidth: 220 }}
                labelId="sorting-selector-label"
                id="sorting-selector"
                label={t("chartSettings.sort")}
                value={visualizationSettings.sorting}
                onChange={(event) => setVisualizationSettingsUserInput({ ...visualizationSettings, sorting: event.target.value })}
            >
                {sortingOptions.map(v => { return <MenuItem key={"key" + v.code} value={v.code}>{v.description[uiContentLanguage]}</MenuItem> })}
            </Select>
        </FormControl>
    );
}

export default SortingSelector;