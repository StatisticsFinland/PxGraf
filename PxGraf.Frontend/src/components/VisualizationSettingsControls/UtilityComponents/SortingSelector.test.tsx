import { render } from '@testing-library/react';
import { ISortingOption } from "../../../types/editorContentsResponse";
import { IVisualizationSettings } from "../../../types/visualizationSettings";
import UiLanguageContext from '../../../contexts/uiLanguageContext';
import { EditorContext } from '../../../contexts/editorContext';
import React from 'react';
import SortingSelector from './SortingSelector';
import userEvent from '@testing-library/user-event';

const visualizationSettings: IVisualizationSettings = {
    sorting: "foo"
};

const sortingOptions: ISortingOption[] = [
    {
        code: "foo",
        description: {
            fi: "foo.fi",
            sv: "foo.sv",
            en: "foo.en"
        }
    },
    {
        code: "bar",
        description: {
            fi: "bar.fi",
            sv: "bar.sv",
            en: "bar.en"
        }
    }
];

const languageContext =
{
    language: 'fi',
    setLanguage: jest.fn(),
    languageTab: 'fi',
    setLanguageTab: jest.fn(),
    availableUiLanguages: ['fi', 'en', 'sv'],
    uiContentLanguage: 'fi',
    setUiContentLanguage: jest.fn()
};

const editorContext = {
    cubeQuery: null,
    setCubeQuery: jest.fn(),
    query: null,
    setQuery: jest.fn(),
    saveDialogOpen: false,
    setSaveDialogOpen: jest.fn(),
    selectedVisualizationUserInput: null,
    setSelectedVisualizationUserInput: jest.fn(),
    visualizationSettingsUserInput: null,
    setVisualizationSettingsUserInput: jest.fn(),
    defaultSelectables: null,
    setDefaultSelectables: jest.fn()
};

describe('rendering test', () => {
    it('renders sorting selector correctly', () => {
        const { asFragment } = render(
            <UiLanguageContext.Provider value={languageContext}>
                <EditorContext.Provider value={editorContext}>
                    <SortingSelector visualizationSettings={visualizationSettings} sortingOptions={sortingOptions} />
                </EditorContext.Provider>
            </UiLanguageContext.Provider>
        );

        expect(asFragment()).toMatchSnapshot();
    })
})

describe('functionality test', () => {
    it('calls setVisualizationSettingsUserInput with the correct value when value is selected', async () => {
        const user = userEvent.setup();
        const mockSetVisualizationSettingsUserInput = jest.fn();
        const { getByLabelText, getByText } = render(
            <UiLanguageContext.Provider value={languageContext}>
                <EditorContext.Provider value={{ ...editorContext, setVisualizationSettingsUserInput: mockSetVisualizationSettingsUserInput }}>
                    <SortingSelector visualizationSettings={visualizationSettings} sortingOptions={sortingOptions} />
                </EditorContext.Provider>
            </UiLanguageContext.Provider>
        );
        const sortingSelector = getByLabelText('chartSettings.sort');
        await user.click(sortingSelector);
        const option = getByText('bar.fi');
        await user.click(option);
        expect(mockSetVisualizationSettingsUserInput).toHaveBeenCalledWith({
            ...visualizationSettings,
            sorting: 'bar'
        });
    });
});