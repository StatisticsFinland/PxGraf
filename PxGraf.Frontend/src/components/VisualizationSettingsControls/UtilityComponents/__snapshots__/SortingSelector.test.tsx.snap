// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rendering test renders sorting selector correctly 1`] = `
<DocumentFragment>
  <div
    class="MuiFormControl-root css-1iw3t7y-MuiFormControl-root"
  >
    <label
      class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiFormLabel-filled MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined css-113d811-MuiFormLabel-root-MuiInputLabel-root"
      data-shrink="true"
      id="sorting-selector-label"
    >
      chartSettings.sort
    </label>
    <div
      class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiSelect-root css-j5ihgt-MuiInputBase-root-MuiOutlinedInput-root-MuiSelect-root"
    >
      <div
        aria-expanded="false"
        aria-haspopup="listbox"
        aria-labelledby="sorting-selector-label sorting-selector"
        class="MuiSelect-select MuiSelect-outlined MuiInputBase-input MuiOutlinedInput-input css-w76bbz-MuiSelect-select-MuiInputBase-input-MuiOutlinedInput-input"
        id="sorting-selector"
        role="combobox"
        tabindex="0"
      >
        foo.fi
      </div>
      <input
        aria-hidden="true"
        aria-invalid="false"
        class="MuiSelect-nativeInput css-j0riat-MuiSelect-nativeInput"
        tabindex="-1"
        value="foo"
      />
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiSelect-iconOutlined css-lohd6h-MuiSvgIcon-root-MuiSelect-icon"
        data-testid="ArrowDropDownIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M7 10l5 5 5-5z"
        />
      </svg>
      <fieldset
        aria-hidden="true"
        class="MuiOutlinedInput-notchedOutline css-1ll44ll-MuiOutlinedInput-notchedOutline"
      >
        <legend
          class="css-w1u3ce"
        >
          <span>
            chartSettings.sort
          </span>
        </legend>
      </fieldset>
    </div>
  </div>
</DocumentFragment>
`;
